### PR TITLE
Binary tree/testing improvements

### PR DESCRIPTION
**Pull request by eloylp**
Hello @mmontes11 , as we talked, here are the changes in the generics POC tree. I think the commits are atomic enough to do the review one by one. I executed some benchmarks after implementing all the changes. The benchmarks were run under the following conditions:

* My laptop was with power plugged in.
* My laptop was configured for "performance" in terms of energy management.
* Checked that no other CPU intensive tasks were happening while running tests.

In order to try to detect weird behaviours, I executed the benchmarking 3 times per each element we are testing (normal tree and its generic implementation).

Here are the results:

#### Try 1
```bash
goos: linux
goarch: amd64
pkg: github.com/coruna-gophers/generics-poc/algo/tree
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkInsert
BenchmarkInsert/BenchmarkInsert_100
BenchmarkInsert/BenchmarkInsert_100-8              47178            124839 ns/op              56 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000
BenchmarkInsert/BenchmarkInsert_1000-8             46220            124440 ns/op              72 B/op          4 allocs/op
BenchmarkInsert/BenchmarkInsert_10000
BenchmarkInsert/BenchmarkInsert_10000-8            47412            126513 ns/op              72 B/op          4 allocs/op
BenchmarkInsert/BenchmarkInsert_1000000
BenchmarkInsert/BenchmarkInsert_1000000-8          48063            146424 ns/op              72 B/op          4 allocs/op
BenchmarkFind
BenchmarkFind/BenchmarkFind_100
BenchmarkFind/BenchmarkFind_100-8               70419894                15.79 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000
BenchmarkFind/BenchmarkFind_1000-8              37899154                43.37 ns/op            8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_10000
BenchmarkFind/BenchmarkFind_10000-8             39667114                59.12 ns/op            8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_1000000
BenchmarkFind/BenchmarkFind_1000000-8           25700118                43.56 ns/op            8 B/op          1 allocs/op
PASS
ok      github.com/coruna-gophers/generics-poc/algo/tree        50.958s
```
#### Try 2
```bash
goos: linux
goarch: amd64
pkg: github.com/coruna-gophers/generics-poc/algo/tree
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkInsert
BenchmarkInsert/BenchmarkInsert_100
BenchmarkInsert/BenchmarkInsert_100-8              45918            122720 ns/op              56 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000
BenchmarkInsert/BenchmarkInsert_1000-8             46803            125815 ns/op              72 B/op          4 allocs/op
BenchmarkInsert/BenchmarkInsert_10000
BenchmarkInsert/BenchmarkInsert_10000-8            45979            136015 ns/op              72 B/op          4 allocs/op
BenchmarkInsert/BenchmarkInsert_1000000
BenchmarkInsert/BenchmarkInsert_1000000-8          44509            143761 ns/op              72 B/op          4 allocs/op
BenchmarkFind
BenchmarkFind/BenchmarkFind_100
BenchmarkFind/BenchmarkFind_100-8               73872097                15.53 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000
BenchmarkFind/BenchmarkFind_1000-8              40037649                43.08 ns/op            8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_10000
BenchmarkFind/BenchmarkFind_10000-8             39854592                58.76 ns/op            8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_1000000
BenchmarkFind/BenchmarkFind_1000000-8           25864032                43.34 ns/op            8 B/op          1 allocs/op
PASS
ok      github.com/coruna-gophers/generics-poc/algo/tree        50.903s
````

#### Try 3
```bash
goos: linux
goarch: amd64
pkg: github.com/coruna-gophers/generics-poc/algo/tree
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkInsert
BenchmarkInsert/BenchmarkInsert_100
BenchmarkInsert/BenchmarkInsert_100-8              46779            123766 ns/op              56 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000
BenchmarkInsert/BenchmarkInsert_1000-8             45399            121567 ns/op              72 B/op          4 allocs/op
BenchmarkInsert/BenchmarkInsert_10000
BenchmarkInsert/BenchmarkInsert_10000-8            47199            127016 ns/op              72 B/op          4 allocs/op
BenchmarkInsert/BenchmarkInsert_1000000
BenchmarkInsert/BenchmarkInsert_1000000-8          45878            142849 ns/op              72 B/op          4 allocs/op
BenchmarkFind
BenchmarkFind/BenchmarkFind_100
BenchmarkFind/BenchmarkFind_100-8               71865142                15.87 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000
BenchmarkFind/BenchmarkFind_1000-8              39938185                43.60 ns/op            8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_10000
BenchmarkFind/BenchmarkFind_10000-8             40040041                58.63 ns/op            8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_1000000
BenchmarkFind/BenchmarkFind_1000000-8           26114683                42.87 ns/op            8 B/op          1 allocs/op
PASS
ok      github.com/coruna-gophers/generics-poc/algo/tree        50.205s

```

#### Try 1
```bash
goos: linux
goarch: amd64
pkg: github.com/coruna-gophers/generics-poc/algo/gtree
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkInsert
BenchmarkInsert/BenchmarkInsert_100
BenchmarkInsert/BenchmarkInsert_100-8              53024            121651 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000
BenchmarkInsert/BenchmarkInsert_1000-8             55668            128026 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_10000
BenchmarkInsert/BenchmarkInsert_10000-8            48220            126181 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000000
BenchmarkInsert/BenchmarkInsert_1000000-8          52323            133608 ns/op              40 B/op          2 allocs/op
BenchmarkFind
BenchmarkFind/BenchmarkFind_100
BenchmarkFind/BenchmarkFind_100-8               87725556                12.39 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000
BenchmarkFind/BenchmarkFind_1000-8              75512365                24.71 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_10000
BenchmarkFind/BenchmarkFind_10000-8             78239480                34.68 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000000
BenchmarkFind/BenchmarkFind_1000000-8           54656458                46.92 ns/op            0 B/op          0 allocs/op
PASS
ok      github.com/coruna-gophers/generics-poc/algo/gtree       54.058s
````
#### Try 2

```bash
goos: linux
goarch: amd64
pkg: github.com/coruna-gophers/generics-poc/algo/gtree
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkInsert
BenchmarkInsert/BenchmarkInsert_100
BenchmarkInsert/BenchmarkInsert_100-8              55288            125846 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000
BenchmarkInsert/BenchmarkInsert_1000-8             56583            129378 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_10000
BenchmarkInsert/BenchmarkInsert_10000-8            55228            145868 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000000
BenchmarkInsert/BenchmarkInsert_1000000-8          53536            138767 ns/op              40 B/op          2 allocs/op
BenchmarkFind
BenchmarkFind/BenchmarkFind_100
BenchmarkFind/BenchmarkFind_100-8               91231912                12.48 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000
BenchmarkFind/BenchmarkFind_1000-8              78649864                25.32 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_10000
BenchmarkFind/BenchmarkFind_10000-8             75481576                34.42 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000000
BenchmarkFind/BenchmarkFind_1000000-8           54214054                52.37 ns/op            0 B/op          0 allocs/op
PASS
ok      github.com/coruna-gophers/generics-poc/algo/gtree       57.584s
````

#### Try 3
```bash
goos: linux
goarch: amd64
pkg: github.com/coruna-gophers/generics-poc/algo/gtree
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkInsert
BenchmarkInsert/BenchmarkInsert_100
BenchmarkInsert/BenchmarkInsert_100-8              54890            140897 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000
BenchmarkInsert/BenchmarkInsert_1000-8             49756            113391 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_10000
BenchmarkInsert/BenchmarkInsert_10000-8            55574            144149 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000000
BenchmarkInsert/BenchmarkInsert_1000000-8          53466            138500 ns/op              40 B/op          2 allocs/op
BenchmarkFind
BenchmarkFind/BenchmarkFind_100
BenchmarkFind/BenchmarkFind_100-8               89916878                12.37 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000
BenchmarkFind/BenchmarkFind_1000-8              80274172                24.50 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_10000
BenchmarkFind/BenchmarkFind_10000-8             78942948                34.72 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000000
BenchmarkFind/BenchmarkFind_1000000-8           44104239                54.95 ns/op            0 B/op          0 allocs/op
PASS
ok      github.com/coruna-gophers/generics-poc/algo/gtree       57.668s
```

### Observations and questions

* We can clearly see how the generics implementation skips 2 allocations. Could the key comparator function be the guilty of that ? it is doing 2 extra type assertions. However, we still need to check this in the profile.
* Even skipping that 2 allocations, i cant appreciate BIG differences in terms of performance among the 2 implementations. In some cases, looks like the non generic implementation performs better ! 
* It is possible that this results are lying to us - We generate the trees dataset with the function `helper.GenerateRandomSliceSet()` . As we generate different random datasets per each tree test, theres a possibility of unfairness. Probably generating this dataset once (i.e save to a  file and parse later), ensuring its the same, balanced dataset for both implementations, would be a better test ? Another option would be to add a `rebalance` feature to both trees, and invoke it before the benchmark.

This is are some **initial observations** . Please, take a look whenever you can with your x-ray vision :) .

### CODE REVIEW

**Comment by eloylp** [comment]
We missed to prevent some [compiler optimizations](https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go):

> Before concluding I wanted to highlight that to be completely accurate, any benchmark should be careful to avoid compiler optimisations eliminating the function under test and artificially lowering the run time of the benchmark.
```go
var result int

func BenchmarkFibComplete(b *testing.B) {
        var r int
        for n := 0; n < b.N; n++ {
                // always record the result of Fib to prevent
                // the compiler eliminating the function call.
                r = Fib(10)
        }
        // always store the result to a package level variable
        // so the compiler cannot eliminate the Benchmark itself.
        result = r
}
```
fixed at https://github.com/coruna-gophers/generics-poc/pull/1/commits/26e4081361999c4b983c3f0778b05897c7498cf7

Here are the new, totally different results:

### Non generic tree
#### Try 1
```bash
goos: linux
goarch: amd64
pkg: github.com/coruna-gophers/generics-poc/algo/tree
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkInsert
BenchmarkInsert/BenchmarkInsert_100
BenchmarkInsert/BenchmarkInsert_100-8              15213            121521 ns/op              56 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000
BenchmarkInsert/BenchmarkInsert_1000-8             15024            127148 ns/op              72 B/op          4 allocs/op
BenchmarkInsert/BenchmarkInsert_10000
BenchmarkInsert/BenchmarkInsert_10000-8            15010            118676 ns/op              72 B/op          4 allocs/op
BenchmarkInsert/BenchmarkInsert_1000000
BenchmarkInsert/BenchmarkInsert_1000000-8          15338            122930 ns/op              72 B/op          4 allocs/op
BenchmarkFind
BenchmarkFind/BenchmarkFind_100
BenchmarkFind/BenchmarkFind_100-8               25151128                41.99 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000
BenchmarkFind/BenchmarkFind_1000-8              15207322               118.9 ns/op             8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_10000
BenchmarkFind/BenchmarkFind_10000-8             14564948               156.6 ns/op             8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_1000000
BenchmarkFind/BenchmarkFind_1000000-8            8774126               126.1 ns/op             8 B/op          1 allocs/op
PASS
ok      github.com/coruna-gophers/generics-poc/algo/tree        74.363s
```
#### Try 2
```bash
goos: linux
goarch: amd64
pkg: github.com/coruna-gophers/generics-poc/algo/tree
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkInsert
BenchmarkInsert/BenchmarkInsert_100
BenchmarkInsert/BenchmarkInsert_100-8              14246            119850 ns/op              56 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000
BenchmarkInsert/BenchmarkInsert_1000-8             15259            122191 ns/op              72 B/op          4 allocs/op
BenchmarkInsert/BenchmarkInsert_10000
BenchmarkInsert/BenchmarkInsert_10000-8            15340            124636 ns/op              72 B/op          4 allocs/op
BenchmarkInsert/BenchmarkInsert_1000000
BenchmarkInsert/BenchmarkInsert_1000000-8          15412            125859 ns/op              72 B/op          4 allocs/op
BenchmarkFind
BenchmarkFind/BenchmarkFind_100
BenchmarkFind/BenchmarkFind_100-8               28274516                62.00 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000
BenchmarkFind/BenchmarkFind_1000-8              14882094               121.4 ns/op             8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_10000
BenchmarkFind/BenchmarkFind_10000-8             14594961               165.2 ns/op             8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_1000000
BenchmarkFind/BenchmarkFind_1000000-8            8655182               115.9 ns/op             8 B/op          1 allocs/op
PASS
ok      github.com/coruna-gophers/generics-poc/algo/tree        77.749s
```
#### Try 3
```bash
goos: linux
goarch: amd64
pkg: github.com/coruna-gophers/generics-poc/algo/tree
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkInsert
BenchmarkInsert/BenchmarkInsert_100
BenchmarkInsert/BenchmarkInsert_100-8              14970            119927 ns/op              56 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000
BenchmarkInsert/BenchmarkInsert_1000-8             15253            123022 ns/op              72 B/op          4 allocs/op
BenchmarkInsert/BenchmarkInsert_10000
BenchmarkInsert/BenchmarkInsert_10000-8            14772            119085 ns/op              72 B/op          4 allocs/op
BenchmarkInsert/BenchmarkInsert_1000000
BenchmarkInsert/BenchmarkInsert_1000000-8          15122            121810 ns/op              72 B/op          4 allocs/op
BenchmarkFind
BenchmarkFind/BenchmarkFind_100
BenchmarkFind/BenchmarkFind_100-8               24371120                43.20 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000
BenchmarkFind/BenchmarkFind_1000-8              14760550               117.2 ns/op             8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_10000
BenchmarkFind/BenchmarkFind_10000-8             12381583               160.1 ns/op             8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_1000000
BenchmarkFind/BenchmarkFind_1000000-8            9030291               112.3 ns/op             8 B/op          1 allocs/op
PASS
ok      github.com/coruna-gophers/generics-poc/algo/tree        74.608s
```

### Generic tree
#### Try 1
```bash
goos: linux
goarch: amd64
pkg: github.com/coruna-gophers/generics-poc/algo/gtree
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkInsert
BenchmarkInsert/BenchmarkInsert_100
BenchmarkInsert/BenchmarkInsert_100-8              17673            122386 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000
BenchmarkInsert/BenchmarkInsert_1000-8             17826            122440 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_10000
BenchmarkInsert/BenchmarkInsert_10000-8            17973            122147 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000000
BenchmarkInsert/BenchmarkInsert_1000000-8          18178            121296 ns/op              40 B/op          2 allocs/op
BenchmarkFind
BenchmarkFind/BenchmarkFind_100
BenchmarkFind/BenchmarkFind_100-8               27408546                38.37 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000
BenchmarkFind/BenchmarkFind_1000-8              17200729                96.22 ns/op            8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_10000
BenchmarkFind/BenchmarkFind_10000-8             17038800               124.6 ns/op             8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_1000000
BenchmarkFind/BenchmarkFind_1000000-8           10474894                98.32 ns/op            8 B/op          1 allocs/op
PASS
ok      github.com/coruna-gophers/generics-poc/algo/gtree       63.974s
```
#### Try 2
```bash
goos: linux
goarch: amd64
pkg: github.com/coruna-gophers/generics-poc/algo/gtree
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkInsert
BenchmarkInsert/BenchmarkInsert_100
BenchmarkInsert/BenchmarkInsert_100-8              17042            115695 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000
BenchmarkInsert/BenchmarkInsert_1000-8             18018            122762 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_10000
BenchmarkInsert/BenchmarkInsert_10000-8            17857            121182 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000000
BenchmarkInsert/BenchmarkInsert_1000000-8          18385            122877 ns/op              40 B/op          2 allocs/op
BenchmarkFind
BenchmarkFind/BenchmarkFind_100
BenchmarkFind/BenchmarkFind_100-8               30264703                36.43 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000
BenchmarkFind/BenchmarkFind_1000-8              15578973                95.53 ns/op            8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_10000
BenchmarkFind/BenchmarkFind_10000-8             16156694               122.6 ns/op             8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_1000000
BenchmarkFind/BenchmarkFind_1000000-8           10495038                97.74 ns/op            8 B/op          1 allocs/op
PASS
ok      github.com/coruna-gophers/generics-poc/algo/gtree       63.159s
```
#### Try 3
```bash
goos: linux
goarch: amd64
pkg: github.com/coruna-gophers/generics-poc/algo/gtree
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkInsert
BenchmarkInsert/BenchmarkInsert_100
BenchmarkInsert/BenchmarkInsert_100-8              17557            122427 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000
BenchmarkInsert/BenchmarkInsert_1000-8             17198            131488 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_10000
BenchmarkInsert/BenchmarkInsert_10000-8            17536            131198 ns/op              40 B/op          2 allocs/op
BenchmarkInsert/BenchmarkInsert_1000000
BenchmarkInsert/BenchmarkInsert_1000000-8          17275            130503 ns/op              40 B/op          2 allocs/op
BenchmarkFind
BenchmarkFind/BenchmarkFind_100
BenchmarkFind/BenchmarkFind_100-8               30087495                44.07 ns/op            0 B/op          0 allocs/op
BenchmarkFind/BenchmarkFind_1000
BenchmarkFind/BenchmarkFind_1000-8              14217362               107.3 ns/op             8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_10000
BenchmarkFind/BenchmarkFind_10000-8             15844375               138.8 ns/op             8 B/op          1 allocs/op
BenchmarkFind/BenchmarkFind_1000000
BenchmarkFind/BenchmarkFind_1000000-8           12272775               194.3 ns/op             8 B/op          1 allocs/op
PASS
ok      github.com/coruna-gophers/generics-poc/algo/gtree       74.365s

```

File: algo/gtree/tree.go [hunk 1]
```diff
@@ -42,7 +42,7 @@ func (t *Tree[K, V]) Insert(key K, value V) {
 }
 
 func (t *Tree[K, V]) insert(n *node[K, V], entry *entry[K, V]) {
-	c := t.compare(n.entry.key, entry.key)
```
Comment (mmontes11) [inline algo/gtree/tree.go:45]: Nice catch 🎖️

File: algo/gtree/tree_test.go [hunk 2]
```diff
@@ -73,58 +73,51 @@ func TestTree_FindNotFound(t *testing.T) {
...
-			b.Log("after getree")
-
-			key := randRange(0, n)
-			b.Log("after rand")
 
 			b.ResetTimer()
-			b.Log("after timer")
 			for i := 0; i < b.N; i++ {
-				tr.Find(key)
+				tr.Insert(key, key)
```
Comment (mmontes11) [inline algo/gtree/tree_test.go:87]: The problem I see with this approach is that we are reusing the same `tr` instance `b.N` times:
- Every iteration mutates the `tr` instance adding new elements
- The `key` variable is not updated with the latest `tr` state

My point here is that `key` would be only the worst case in the first iteration, but not in the rest. Maybe I am missing something here?

Anyway, for having an estimation it was more than fair scenario.

File: algo/gtree/tree_test.go [hunk 3]
```diff
@@ -73,58 +73,51 @@ func TestTree_FindNotFound(t *testing.T) {
...
+			s := helper.GenerateRandomSliceSet(n)
+			key := helper.Higher(s)
+			tr := getTree(s)
+
+			var o1 interface{}
+			var o2 bool
+
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				o1, o2 = tr.Find(key)
```
Comment (mmontes11) [inline algo/gtree/tree_test.go:108]: Thank you @davecheney, very inspiring [article](https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go)!

**Review by mmontes11 — APPROVED** [review]
Thank you for this in-depth analysis @eloylp !

I've raised some minor points but I cannot agree more with the conclusions.

> We can clearly see how the generics implementation skips 2 allocations. Could the key comparator function be the guilty of that ? it is doing 2 extra type assertions. However, we still need to check this in the profile.

Honestly I would like to understand how a type assertion can incur in extra allocations? Probably we will need to check Golang internals to have a better context.

> Even skipping that 2 allocations, i cant appreciate BIG differences in terms of performance among the 2 implementations. In some cases, looks like the non generic implementation performs better !

Agreed, performance shouldn't be the reason of choice, but mantainability and readability should instead,


> It is possible that this results are lying to us - We generate the trees dataset with the function helper.GenerateRandomSliceSet() . As we generate different random datasets per each tree test, theres a possibility of unfairness.

Actually I think every insertion is performed in a unfair scenario, check the comments in case I'm missing something.

File: algo/gtree/tree_test.go (hunk 2 above)
Comment (eloylp) [reply algo/gtree/tree_test.go:87]: Yes i think you are totally right :+1:  , i couldn't find better solution than this :sweat: Maybe we can take a look at this again on the next iteration :)

File: algo/gtree/tree_test.go (hunk 3 above)
Comment (eloylp) [reply algo/gtree/tree_test.go:108]: Absolutely, thank you !

**Comment by eloylp** [comment]
Okay , seems this PR is ready to merge. Its true that probably we did not reach the best solution and conclusions, but under my point of view, we are closer than before. Next iteration will be better. Merging !

**Pull request merged**